File: Binary-Sreach-1D/Sreachpostion.java
/*
 * Program: Search Insert Position using Binary Search
 * This program finds the index at which a target value K 
 * should be inserted in a sorted array so that the array remains sorted.
 * If K already exists in the array, it returns the index of K.
 */
import java.util.Scanner;
public class Sreachpostion
{
    // Method to find the index to insert K using Binary Search
    public static int searchInsertK(int arr[], int k)
    {
        // Initialize the search boundaries
        int low = 0;
        int high = arr.length - 1;
        while (low <= high)
        {
            // Calculate the middle index to avoid overflow
            int mid = low + (high - low) / 2;
            // If K is found at mid, return the index
            if (arr[mid] == k)
            {
                return mid;
            }
            // If the middle element is smaller than K, search the right half
            else if (arr[mid] < k)
            {
                low = mid + 1;
            }
            // If the middle element is greater than K, search the left half
            else
            {
                high = mid - 1;
            }
        }
        // If not found, low will be the correct position to insert K
        return low; // Return the position where K can be inserted
    }
    public static void main(String[] args)
    {
        // Taking input from the user
        Scanner sc = new Scanner(System.in);
        // Ask user for the size of the array
        System.out.print("Enter the size of the array: ");
        int n = sc.nextInt();
        // Create an array of given size
        int arr[] = new int[n];
        // Take array elements as input (must be sorted)
        System.out.println("Enter the elements of the array in sorted order:");
        for (int i = 0; i < n; i++)
        {
            arr[i] = sc.nextInt();
        }
        // Ask user for the target value K
        System.out.print("Enter the value of K to search: ");
        int k = sc.nextInt();
        // Call the method to find the position of K
        int position = searchInsertK(arr, k);
        // Print the result
        System.out.println("The position to insert K is: " + position);
    }
}

File: Binary-Sreach-1D/Sreachpostion.py
# WAP to find the sreach insertion postion in the array
# leetcode :-https://leetcode.com/problems/search-insert-position/
# geeks for geeks:https://www.geeksforgeeks.org/problems/search-insert-position-of-k-in-a-sorted-array/1
# code360:-https://www.naukri.com/code360/problems/algorithm-to-find-best-insert-position-in-sorted-array_839813?leftPanelTabValue=SUBMISSION

# find or inset in the array 
def searchInsertK(arr, k):
    # assigned the left and right pointer to the first and last index of the array
    left = 0
    right = len(arr) - 1

    # loop until the left pointer is less than or equal to the right pointer
    while left <= right:
        mid = (left + right) // 2

        if arr[mid] == k:
            return mid
        elif arr[mid] < k:
            left = mid + 1
        else:
            right = mid - 1

    return left
                






arr=[]
num=int(input(" Enter the size of array :- "))
print(" Enter the elements ")
for i in range(num):
    i=int(input())
    arr.append(i)
k=int(input("Enter the number to insert sreach postion"))
result=searchInsertK(arr, k)
print("Position of the number : " , searchInsertK(arr, k))